lib/puppet/provider/archive/ruby.rb

```diff
@@ -201,12 +202,13 @@ def transfer_download(archive_filepath)
       actual_checksum = archive.checksum(resource[:checksum_type])
       if actual_checksum != checksum
         destroy
-        FileUtils.rm_f(temppath) if File.exist?(temppath)
         raise(Puppet::Error, "Download file checksum mismatch (expected: #{checksum} actual: #{actual_checksum})")
       end
     end
 
     move_file_in_place(temppath, archive_filepath)
+  ensure
+    FileUtils.rm_f(temppath) if File.exist?(temppath)
```

Comment:
Technically you can rm_f a file that does not exist (because force is only about ignoring errors). Checking for existence does not hurt though, so I am fine with this.